### PR TITLE
stdenv: start deprecating non-list configureFlags

### PR DESCRIPTION
**Pull request by Artturin**
the motivation for this is to simplify stdenv and ease the job of
reviewers due to them needing to tell contributors about the defacto
rule that configureFlags should be a list of strings

Supersedes https://github.com/NixOS/nixpkgs/pull/45886

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by Ericson2314 — APPROVED** [review]
Woo!!

**Review by Mindavi — APPROVED** [review]

**Comment by Artturin** [comment]
added a release note

File: nixos/doc/manual/release-notes/rl-2205.section.md [hunk 1]
```diff
@@ -193,6 +193,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 - `security.klogd` was removed.  Logging of kernel messages is handled
   by systemd since Linux 3.5.
 
+- String 'configureFlags' is deprecated and will be removed in release 23.05. Please use a list of strings.
+
```
Comment (roberth) [inline nixos/doc/manual/release-notes/rl-2205.section.md:197]: It's good to document the change, but it seems noisy to bother readers with changes that they are equally well informed about by the evaluation warnings.

A "safe to skip" section would be nice.

File: pkgs/stdenv/generic/make-derivation.nix [hunk 2]
```diff
@@ -331,8 +331,8 @@ else let
 
       # This parameter is sometimes a string, sometimes null, and sometimes a list, yuck
       configureFlags = let inherit (lib) optional elem; in
-        (/**/ if lib.isString configureFlags then [configureFlags]
-         else if configureFlags == null      then []
+        (/**/ if lib.isString configureFlags then lib.warn "String 'configureFlags' is deprecated and will be removed in release 23.05. Please use a list of strings" [configureFlags]
+         else if configureFlags == null      then lib.warn "Null 'configureFlags' is deprecated and will be removed in release 23.05. Please use a empty list instead '[]'." []
```
Comment (roberth) [inline pkgs/stdenv/generic/make-derivation.nix:335]: Please add derivation name and, if available, source position (`pos` / `meta.position`).

File: nixos/doc/manual/release-notes/rl-2205.section.md (hunk 1 above)
Comment (Artturin) [reply nixos/doc/manual/release-notes/rl-2205.section.md]: removed them

did it with `git checkout @~3 -- ./nixos/doc/manual/release-notes/rl-2205.section.md` to both the md and xml and then commited and fix'd up the commit. someone may find it useful so im posting it

File: pkgs/stdenv/generic/make-derivation.nix (hunk 2 above)
Comment (Artturin) [reply pkgs/stdenv/generic/make-derivation.nix]: i added `pos.file`

**Review by SuperSandro2000 — APPROVED** [review]
🎉

File: pkgs/stdenv/generic/make-derivation.nix [hunk 3]
```diff
@@ -331,8 +331,8 @@ else let
 
       # This parameter is sometimes a string, sometimes null, and sometimes a list, yuck
       configureFlags = let inherit (lib) optional elem; in
-        (/**/ if lib.isString configureFlags then [configureFlags]
-         else if configureFlags == null      then []
+        (/**/ if lib.isString configureFlags then lib.warn "String 'configureFlags' is deprecated and will be removed in release 23.05. Please use a list of strings. ${pos.file}" [configureFlags]
+         else if configureFlags == null      then lib.warn "Null 'configureFlags' is deprecated and will be removed in release 23.05. Please use a empty list instead '[]'. ${pos.file}" []
```
Comment (roberth) [inline pkgs/stdenv/generic/make-derivation.nix:335]:
```suggestion
        (/**/ if lib.isString configureFlags then lib.warn "String 'configureFlags' is deprecated and will be removed in release 23.05. Please use a list of strings. Derivation name: ${derivationArg.name }, file: ${pos.file or "unknown file"}" [configureFlags]
         else if configureFlags == null      then lib.warn "Null 'configureFlags' is deprecated and will be removed in release 23.05. Please use a empty list instead '[]'. Derivation name: ${derivationArg.name}, file: ${pos.file or "unknown file"" []
```

^ we can't assume a non-null `pos`.

also, having the name may be helpful.

Comment (Artturin) [reply pkgs/stdenv/generic/make-derivation.nix]: thanks

**Review by roberth — APPROVED** [review]

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/180067

**Pull request merged**